lib/msal-core/src/messaging/MessageListener.ts

```diff
@@ -0,0 +1,113 @@
+        this.logger = logger;
+        this.topFrameOrigin = topFrameOrigin;
+        this.embeddedFrameOrigin = embeddedFrameOrigin;
+
+        // listen to all incoming messages
+        window.addEventListener("message", this.receiveMessage, false);
+    }
+
+    public setCallBack(consentNeeded: boolean, callback: iframeRedirectCallback) {
+        this.consentNeeded = consentNeeded;
+        this.iframeRedirectCallback = callback;
+    }
+
+    /**
+     * top frame application calls this function to ack to proceed
+     * @param message
+     */
+    processIframeRedirectCallback(url: string) {
+        return () => WindowUtils.navigateWindow(url, this.logger);
+    }
+
+    /**
+     * Parses the messages receieved
+     * This will be a unique handler per message, we will allow only one active request at a time
+     * @param event
+     */
+    private receiveMessage(event: any) {
+
+        const windowType = MessageHelper.currentWindow();
+        const receivedMessage: PAYLOAD = { ...event.data};
+
+        switch(windowType) {
+
+            // Top framed application: handles the delegation on behalf of the iframed app
+            case WindowType.TOP_FRAME: {
+
+                switch(receivedMessage.type) {
+
+                    case MessageType.REDIRECT_REQUEST: {
+                        // acknowlege the redirect on behalf of the iframed app by sending the current location
+                        const message = MessageHelper.buildMessage(MessageType.URL_TOP_FRAME, window.location.href);
+                        MessageDispatcher.dispatchMessage(event.source, message, this.embeddedFrameOrigin);
+                        break;
+                    }
+
+                    case MessageType.URL_NAVIGATE: {
+                        // if the response is the URL to navigate for token acquisition, navigate to STS
+                        if (receivedMessage.type === MessageType.URL_NAVIGATE) {
+                            this.logger.info("navigating to the Service on behalf of the iframed app");
+
+                            if(this.consentNeeded) {
+                                this.iframeRedirectCallback(this.processIframeRedirectCallback(receivedMessage.data));
+                            }
+                            else {
+                                WindowUtils.navigateWindow(receivedMessage.data, this.logger);
+                            }
+                        }
+                        break;
+                    }
+                }
+            }
+
+            // embedded (in an iframe) application
+            case WindowType.IFRAME: {
+
+                // check the origin, should match window.top always; message channel may be more secure
+                if (window.top != event.source) {
+                    this.logger.warning("The message origin is not verified");
+                    return;
+                }
+
+                if (receivedMessage.type === MessageType.URL_TOP_FRAME) {
```

Comment:
I might argue a nest switch statement would be good for consistency with the above block.

Reply:
Done